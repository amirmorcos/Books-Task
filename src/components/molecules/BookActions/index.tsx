import Button from "atoms/Button";
import { useAppTheme } from "hooks/useAppTheme";
import React from "react";
import { Image, TouchableOpacity, View } from "react-native";
import { useSafeAreaInsets } from "react-native-safe-area-context";
import styles from "./styles";
import FavoriteIcon from "assets/icons/FavoriteIcon";
import { Images } from "themes/images";

const BookActions = () => {
  const { colors } = useAppTheme();
  const { bottom } = useSafeAreaInsets();
  const themedStyles = styles(colors);

  return (
    <View style={[themedStyles.container, { bottom }]}>
      <TouchableOpacity style={themedStyles.favoriteButton}>
        <Image source={Images.heart} />
      </TouchableOpacity>
      <Button
        overrideContainerStyle={themedStyles.buyButton}
        label="Buy now"
        size="large"
        onPress={() => {}}
      />
    </View>
  );
};

export default BookActions;
