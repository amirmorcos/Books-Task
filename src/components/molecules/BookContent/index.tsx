import { useAppTheme } from "hooks/useAppTheme";
import React from "react";
import { Text, View } from "react-native";
import styles from "./styles";
import { BookContentProps } from "./types";

const BookContent = ({ title, authors }: BookContentProps) => {
  const { colors } = useAppTheme();
  const themedStyles = styles(colors);

  return (
    <View style={themedStyles.container}>
      <Text style={themedStyles.title} numberOfLines={2} ellipsizeMode="tail">
        {title}
      </Text>
      <View style={themedStyles.bottomContainer}>
        {authors?.length && (
          <View style={themedStyles.authorContainer}>
            {authors.slice(0, 2).map((author) => (
              <Text key={author} style={themedStyles.author}>
                {author}
              </Text>
            ))}
          </View>
        )}
      </View>
    </View>
  );
};

export default BookContent;
