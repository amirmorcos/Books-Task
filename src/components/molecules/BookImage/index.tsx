import { useAppTheme } from "hooks/useAppTheme";
import React from "react";
import { Image, View } from "react-native";
import styles from "./styles";
import type { BookImageProps } from "./types";
import FastImage from "react-native-fast-image";
import { fixImageProtocol } from "utils/constants";

const BookImage = ({ image }: BookImageProps) => {
  const { colors } = useAppTheme();
  const themedStyles = styles(colors);
  return (
    <View style={themedStyles.container}>
      <FastImage
        resizeMode="contain"
        source={{ uri: fixImageProtocol(image) }}
        style={themedStyles.image}
      />
    </View>
  );
};

export default BookImage;
