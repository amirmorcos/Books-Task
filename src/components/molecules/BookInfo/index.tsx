import StarIcon from "assets/icons/StarIcon";
import { useAppTheme } from "hooks/useAppTheme";
import React from "react";
import { Text, View } from "react-native";
import styles from "./styles";
import type { BookInfoProps } from "./types";

const BookInfo = ({ categories, pageCount, rating }: BookInfoProps) => {
  const { colors } = useAppTheme();
  const themedStyles = styles(colors);
  return (
    <View style={themedStyles.container}>
      {rating && (
        <View style={themedStyles.ratingContainer}>
          <StarIcon />
          <Text style={themedStyles.rating}>{rating.toFixed(1)}</Text>
        </View>
      )}
      <View style={themedStyles.categoriesContainer}>
        {categories.map((category) => (
          <Text key={category} style={themedStyles.category}>
            {category}
          </Text>
        ))}
      </View>
      <View style={themedStyles.pageContainer}>
        <Text>{pageCount} pages</Text>
      </View>
    </View>
  );
};

export default BookInfo;
