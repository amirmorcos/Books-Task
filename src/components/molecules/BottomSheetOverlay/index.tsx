import React, { forwardRef } from "react";
import {
  BottomSheetModal,
  BottomSheetModalProvider,
  BottomSheetBackdrop,
  BottomSheetScrollView,
} from "@gorhom/bottom-sheet";
import styles from "./styles";
import type { BottomSheetOverlayProps } from "./types";

const BottomSheetOverlay = forwardRef<
  BottomSheetModal,
  BottomSheetOverlayProps
>(({ children, overrideContainerStyle }, ref) => {
  return (
    <BottomSheetModal
      backdropComponent={(backdropProps) => (
        <BottomSheetBackdrop
          {...backdropProps}
          enableTouchThrough
          pressBehavior="close"
        />
      )}
      ref={ref}
      index={1}
      enablePanDownToClose
      enableDynamicSizing
      snapPoints={[1]}
    >
      <BottomSheetScrollView
        showsVerticalScrollIndicator={false}
        enableFooterMarginAdjustment
        style={[styles.container, overrideContainerStyle]}
        contentContainerStyle={styles.content}
        // bounces={false}
      >
        {children}
      </BottomSheetScrollView>
    </BottomSheetModal>
  );
});

export default BottomSheetOverlay;
