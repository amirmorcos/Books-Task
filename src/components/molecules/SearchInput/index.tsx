import { Input } from "atoms/index";
import { useAppTheme } from "hooks/useAppTheme";
import React from "react";
import SearchIcon from "../../../../assets/icons/SearchIcon";
import styles from "./styles";
import { SearchInputProps } from "./types";

const SearchInput = ({
  onSearch,
  overrideContainerStyle,
}: SearchInputProps) => {
  const { colors } = useAppTheme();
  const themedStyles = styles(colors);

  return (
    <Input
      placeholder="Search"
      overrideInputStyle={themedStyles.input}
      overrideContainerStyle={overrideContainerStyle}
      leftIcon={<SearchIcon />}
    />
  );
};

export default SearchInput;
