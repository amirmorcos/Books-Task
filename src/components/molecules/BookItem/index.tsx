import { useAppTheme } from "hooks/useAppTheme";
import BookContent from "molecules/BookContent";
import BookImage from "molecules/BookImage";
import { TouchableOpacity } from "react-native";
import styles from "./styles";
import { BookItemProps } from "./types";

const BookItem = ({ onPress, item }: BookItemProps) => {
  const { colors } = useAppTheme();
  const themedStyles = styles(colors);

  return (
    <TouchableOpacity onPress={onPress} style={themedStyles.container}>
      <BookImage image={item.volumeInfo.imageLinks.thumbnail} />
      <BookContent
        title={item.volumeInfo.title}
        authors={item.volumeInfo.authors}
      />
    </TouchableOpacity>
  );
};

export default BookItem;
