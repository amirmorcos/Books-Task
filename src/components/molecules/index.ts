import SearchInput from "./SearchInput";
import BookItem from "./BookItem";
import BookInfo from "./BookInfo";
import BookActions from "./BookActions";
import BookContent from "./BookContent";
import BookImage from "./BookImage";

export { SearchInput, BookItem, BookInfo, BookActions, BookContent, BookImage };
