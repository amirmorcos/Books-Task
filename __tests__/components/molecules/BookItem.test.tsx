/**
 * @format
 */

import React from "react";

// Note: import explicitly to use the types shipped with jest.

// Note: test renderer must be required after react-native.
import renderer from "react-test-renderer";
import { BookItem } from "molecules/index";

const bookItemMocked = {
  kind: "books#volumes",
  totalItems: 534,
  items: [
    {
      kind: "books#volume",
      id: "sjGWCgAAQBAJ",
      etag: "rlVL/M4/RxI",
      selfLink: "https://www.googleapis.com/books/v1/volumes/sjGWCgAAQBAJ",
      volumeInfo: {
        title: "Mop Hop",
        authors: ["Sharon  Coan"],
        publisher: "Teacher Created Materials",
        publishedDate: "2012-03-01",
        description:
          "Featuring clear text and lively illustrations, beginning readers will learn words that feature rhyming -op sounds such as pop, bop, and hop. The creative storyline and accommodating sight words help readers progress in their early literacy and phonemic skills.",
        industryIdentifiers: [
          {
            type: "ISBN_13",
            identifier: "9781433398919",
          },
          {
            type: "ISBN_10",
            identifier: "1433398915",
          },
        ],
        readingModes: {
          text: false,
          image: true,
        },
        pageCount: 10,
        printType: "BOOK",
        categories: ["Juvenile Fiction"],
        maturityRating: "NOT_MATURE",
        allowAnonLogging: false,
        contentVersion: "0.1.2.0.preview.1",
        panelizationSummary: {
          containsEpubBubbles: false,
          containsImageBubbles: false,
        },
        imageLinks: {
          smallThumbnail:
            "http://books.google.com/books/content?id=sjGWCgAAQBAJ&printsec=frontcover&img=1&zoom=5&edge=curl&source=gbs_api",
          thumbnail:
            "http://books.google.com/books/content?id=sjGWCgAAQBAJ&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api",
        },
        language: "en",
        previewLink:
          "http://books.google.com.eg/books?id=sjGWCgAAQBAJ&printsec=frontcover&dq=Mop&hl=&cd=1&source=gbs_api",
        infoLink:
          "http://books.google.com.eg/books?id=sjGWCgAAQBAJ&dq=Mop&hl=&source=gbs_api",
        canonicalVolumeLink:
          "https://books.google.com/books/about/Mop_Hop.html?hl=&id=sjGWCgAAQBAJ",
      },
      saleInfo: {
        country: "EG",
        saleability: "NOT_FOR_SALE",
        isEbook: false,
      },
      accessInfo: {
        country: "EG",
        viewability: "ALL_PAGES",
        embeddable: true,
        publicDomain: false,
        textToSpeechPermission: "ALLOWED",
        epub: {
          isAvailable: false,
        },
        pdf: {
          isAvailable: true,
          acsTokenLink:
            "http://books.google.com.eg/books/download/Mop_Hop-sample-pdf.acsm?id=sjGWCgAAQBAJ&format=pdf&output=acs4_fulfillment_token&dl_type=sample&source=gbs_api",
        },
        webReaderLink:
          "http://play.google.com/books/reader?id=sjGWCgAAQBAJ&hl=&source=gbs_api",
        accessViewStatus: "SAMPLE",
        quoteSharingAllowed: false,
      },
      searchInfo: {
        textSnippet:
          "Featuring clear text and lively illustrations, beginning readers will learn words that feature rhyming -op sounds such as pop, bop, and hop.",
      },
      userInfo: {},
    },
  ],
};

it("renders correctly", () => {
  const tree = renderer
    .create(<BookItem item={bookItemMocked.items[0]} onPress={jest.fn} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
