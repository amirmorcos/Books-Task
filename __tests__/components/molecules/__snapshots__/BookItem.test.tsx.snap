// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "opacity": 1,
    }
  }
>
  <View>
    <View
      style={
        [
          {
            "overflow": "hidden",
          },
          undefined,
        ]
      }
    >
      <FastImageView
        defaultSource={null}
        resizeMode="contain"
        style={
          {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      />
    </View>
  </View>
  <View>
    <Text
      ellipsizeMode="tail"
      numberOfLines={2}
    >
      Mop Hop
    </Text>
    <View>
      <View>
        <Text>
          Sharon  Coan
        </Text>
      </View>
    </View>
  </View>
</View>
`;
